pkg/controllermanager/controller/project/project_stale_control_test.go

```diff
@@ -0,0 +1,376 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package project_test
+
+import (
+	"context"
+	"time"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	fakeclientmap "github.com/gardener/gardener/pkg/client/kubernetes/clientmap/fake"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	fakeclientset "github.com/gardener/gardener/pkg/client/kubernetes/fake"
+	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
+	. "github.com/gardener/gardener/pkg/controllermanager/controller/project"
+	"github.com/gardener/gardener/pkg/logger"
+	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
+	"github.com/gardener/gardener/pkg/operation/common"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubecoreinformers "k8s.io/client-go/informers"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("ProjectStaleControl", func() {
+	Describe("defaultStaleControl", func() {
+		var (
+			ctrl                         *gomock.Controller
+			k8sGardenRuntimeClient       *mockclient.MockClient
+			k8sGardenRuntimeStatusWriter *mockclient.MockStatusWriter
+			gardenCoreInformerFactory    gardencoreinformers.SharedInformerFactory
+			kubeCoreInformerFactory      kubecoreinformers.SharedInformerFactory
+			clientMap                    *fakeclientmap.ClientMap
+
+			oldNowFunc     func() metav1.Time
+			oldTimeNowFunc func() time.Time
+
+			projectName       = "foo"
+			namespaceName     = "garden-foo"
+			secretName        = "secret"
+			secretBindingName = "secretbinding"
+			quotaName         = "quota"
+
+			minimumLifetimeDays     = 5
+			staleGracePeriodDays    = 10
+			staleExpirationTimeDays = 15
+
+			project       *gardencorev1beta1.Project
+			shoot         *gardencorev1beta1.Shoot
+			plant         *gardencorev1beta1.Plant
+			backupEntry   *gardencorev1beta1.BackupEntry
+			quota         *gardencorev1beta1.Quota
+			secret        *corev1.Secret
+			secretBinding *gardencorev1beta1.SecretBinding
+			cfg           *config.ControllerManagerConfiguration
+
+			control StaleControlInterface
+		)
+
+		BeforeSuite(func() {
+			oldNowFunc = NowFunc
+			oldTimeNowFunc = common.TimeNow
+		})
+
+		AfterSuite(func() {
+			NowFunc = oldNowFunc
```

Comment:
Wouldn't be better if we have the nowFunc as arg of `ReconcileStaleProject` func?


```golang
type StaleControlInterface interface {
	ReconcileStaleProject(project *gardencorev1beta1.Project, nowFunc func() metav1.Time) error
}
```

In this way we could prevent preserving the oldNowFunc in BeforeSuite and setting it back in AfterSuite?

Reply:
ok, sure, thanks, done